src/coreclr/interpreter/compiler.cpp

```diff
@@ -3257,27 +3606,107 @@ int InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
                         m_pLastNewIns->SetDVar(m_pStackPointer[-1].var);
                         m_ip++;
                         break;
+
+                    case CEE_ENDFILTER:
+                        // At this point, we expect the top of the stack is the decision of the filter
+                        // below it is the exception object, and nothing else
+                        assert((m_pStackPointer - m_pStackBase) == 2);
+                        AddIns(INTOP_LEAVE_FILTER);
+                        m_pStackPointer--;
+                        m_pLastNewIns->SetSVar(m_pStackPointer[0].var);
+                        m_ip++;
+                        break;
                     default:
                         assert(0);
                         break;
                 }
                 break;
 
+            case CEE_LEAVE:
+            case CEE_LEAVE_S:
+                {
+                    // At this point, we are leaving the block, we should not have anything on the stack
+                    assert(m_pStackPointer == m_pStackBase);
+                    // We should not have a LEAVE instruction in a filter or a finally block
+                    assert((m_pCBB->funclet_type == TRY_ENTRY) || (m_pCBB->funclet_type == TRY) || (m_pCBB->funclet_type == CATCH_ENTRY) || (m_pCBB->funclet_type == CATCH));
+                    InterpInst* prev = AddIns(INTOP_NOP);
+                    InterpInst* next = AddIns(INTOP_NOP);
+                    int ilOffset = (int)(m_ip - m_pILCode);
+                    int target = (opcode == CEE_LEAVE) ? ilOffset + 5 + *(int32_t*)(m_ip + 1) : (ilOffset + 2 + (int8_t)m_ip[1]);
+                    if (m_pCBB->finally_index != -1)
```

Comment:
A leave opcode can jump over multiple protected blocks. It seems to me that here you only have support for invoking a single finally clause, when it might be the case that you need to invoke more ?

Reply:
This is something I am working on right now. We need to handle "interesting" usage of the leave opcode that exit more than one protected region. This never happens with the C# compiler that generate debug IL and so I wasn't aware of that earlier.

Reply:
This is actually easy to reproduce, for example:
```
Method ()
{
   try {
      try {
         return;
      } finally {
      }
   } finally {
   }
}
```

The return will leave over multiple finally blocks